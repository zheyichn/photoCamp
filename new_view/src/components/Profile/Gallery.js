import { useState, useEffect, useCallback } from "react";
import ImageList from "@mui/material/ImageList";
import { CardMedia } from "@mui/material";
import ImageListItem from "@mui/material/ImageListItem";
import {
  getTagsAPI,
  getUserAPI,
  getPostByIdAPI,
} from "../../api/mock_api_following_post_follower";
import EditPostPopUp from "./EditPostPopUp";
import { Box, Tab, Tabs } from "@mui/material";
import { isImgUrl } from "../../utility/isImageUrl";

export default function Gallery() {
  const [posts, setPosts] = useState([]);
  const [tagPosts, setTagPosts] = useState([]);
  const [, setWhichPost] = useState([]);
  // states to determine whether edit post pop-up is open and what is the current post being edited
  const [editOpen, setEditOpen] = useState(false);
  const [currPost, setCurrPost] = useState("");
  const [tabIndex, setTabIndex] = useState(0);

  const judgeIfImage = async (url) => {
    const res = await isImgUrl(url);
    return res;
  };

  const handleTabChange = (event, newTabIndex) => {
    setTabIndex(newTabIndex);
    if (tabIndex === 0) {
      setWhichPost(tagPosts);
    } else {
      setWhichPost(posts);
    }
  };

  // handle the opening of edit post pop up
  const handleClickOpen = (post) => {
    setEditOpen(true);
    setCurrPost(post);
  };

  // handle the closing of edit post pop up
  const handleClickClose = () => {
    setEditOpen(false);
    setCurrPost("");
  };

  // get one post
  const getOnePost = useCallback(async (postId) => {
    let post = await getPostByIdAPI(postId);
    post = post.data[0];
    const isImage = await judgeIfImage(post.postContent);
    post["isImage"] = isImage;
    return post;
  }, []);

  // get all the posts associated with this user from api
  const getPostContent = useCallback(async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    const postIds = userResponse.data.posts;
    const postsContent = await Promise.all(postIds.map(getOnePost));
    setPosts(postsContent);
  }, [getOnePost]);


  const judgeImageHelper = useCallback(async(post) => {
    const isImage = await judgeIfImage(post.postContent);
    post["isImage"] = isImage;
    return post;
  }, []);
  
  // get tab content associated with the current user
  // and judge whether each post is an image
  const getTagContent = useCallback(async () => {
    const tagResponse = await getTagsAPI(sessionStorage.userName);
    const taggedPosts = tagResponse.data;
    const judgedTaggedPosts = await Promise.all(taggedPosts.map(judgeImageHelper));
    setTagPosts(judgedTaggedPosts);
  }, [judgeImageHelper]);



  // call getPostContent()
  // this should be called initially, AND whenever the pop up opens / closes
  // allow a timeout of 100ms - sometimes we perform a POST request to back end and we are fetching the data again
  useEffect(() => {
    getPostContent();
    getTagContent();
  }, [getPostContent, getTagContent]);

  useEffect(() => {
    setTimeout(() => getPostContent(), 100);
    setTimeout(() => getTagContent(), 100);
    // setTimeout(() => getTagContent(), 100);
  }, [editOpen, getPostContent, getTagContent]);

  return (
    <main>
      <Box>
        <Tabs value={tabIndex} onChange={handleTabChange} centered>
          <Tab label="Posts" />
          <Tab label="Tagged" />
        </Tabs>
      </Box>

      <div className="container">
        {tabIndex ? (
          <div className="gallery">
            <ImageList
              data-testid="img-list"
              sx={{ width: 1000, height: 1000 }}
              cols={3}
              rowHeight={333}
            >
              {tagPosts.map((item) => (
                <ImageListItem key={item._id}>
                  {/* Note: use this when implenting backend for tagging!*/}
                  {item.isImage ? (
                    <CardMedia
                      data-testid="test-img"
                      component="img"
                      height="300px"
                      src={item.postContent}
                      alt="Post Picture"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  ) : (
                    <CardMedia
                      data-testid="test-img"
                      component="video"
                      height="300px"
                      src={item.postContent}
                      autoPlay
                      controls
                      muted
                      alt="Post Video"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  )}
                </ImageListItem>
              ))}
            </ImageList>
          </div>
        ) : (
          <div className="gallery">
            <ImageList
              data-testid="img-list"
              sx={{ width: 1000, height: 1000 }}
              cols={3}
              rowHeight={333}
            >
              {posts.map((item) => (
                <ImageListItem key={item._id}>
                  {item.isImage ? (
                    <CardMedia
                      data-testid="test-img"
                      component="img"
                      height="300px"
                      src={item.postContent}
                      alt="Post Picture"
                      onClick={() => handleClickOpen(item)}
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  ) : (
                    <CardMedia
                      data-testid="test-img"
                      component="video"
                      height="300px"
                      src={item.postContent}
                      autoPlay
                      controls
                      muted
                      onClick={() => handleClickOpen(item)}
                      alt="Post Video"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  )}
                </ImageListItem>
              ))}
            </ImageList>
          </div>
      )}

        <EditPostPopUp
          open={editOpen}
          onClose={handleClickClose}
          currPost={currPost}
        />
      </div>
    </main>
  );
}
