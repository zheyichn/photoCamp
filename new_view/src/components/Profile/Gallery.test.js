import { cleanup , render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import Gallery from "./Gallery";

// afterEach function runs after each test suite is executed
afterEach(() => {
  cleanup();
});

describe("Gallery", () => {
  // Test1: match snapshots
  it("Page renders correctly", () => {
    const tree = renderer.create(<Gallery />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  //Test2: 
  test("image list is displayed", () => {
    render(<Gallery  />);
    const element = screen.getByTestId("img-list");
    expect(element).toBeInTheDocument();

  })
});
