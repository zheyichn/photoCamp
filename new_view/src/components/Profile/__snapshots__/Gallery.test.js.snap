// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery Page renders correctly 1`] = `
<main>
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiTabs-root css-1ujnqem-MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        onScroll={[Function]}
        style={
          {
            "marginBottom": -0,
            "overflow": "hidden",
          }
        }
      >
        <div
          aria-orientation={null}
          className="MuiTabs-flexContainer MuiTabs-centered css-1wf8b0h-MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            Posts
            <span
              className="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style={{}}
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            Tagged
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="gallery"
    >
      <ul
        className="MuiImageList-root MuiImageList-standard css-zpxbs6-MuiImageList-root"
        data-testid="img-list"
        style={
          {
            "gap": 4,
            "gridTemplateColumns": "repeat(3, 1fr)",
          }
        }
      />
    </div>
  </div>
</main>
`;
